File: src/jedimem/store.py
"""Where memories live: the staging ref, and the committed files."""
from __future__ import annotations

import pathlib

from . import repo
from .memory import Memory, MemoryError

MEM_DIR = ".jedimem/memories"
STAGE_PREFIX = "pending/"


class Store:
    def __init__(self, root: pathlib.Path, staging_ref=repo.STAGING_REF):
        self.root = pathlib.Path(root)
        self.ref = staging_ref

    # ------------------------------------------------------------ committed
    @property
    def mem_dir(self) -> pathlib.Path:
        return self.root / MEM_DIR

    def all(self, include_inactive=False) -> list:
        out = []
        if not self.mem_dir.is_dir():
            return out
        for p in sorted(self.mem_dir.glob("*.md")):
            m = Memory.load(p)
            if include_inactive or m.status == "active":
                out.append(m)
        return out

    def by_id(self, mid: str):
        p = self.mem_dir / f"{mid}.md"
        return Memory.load(p) if p.exists() else None

    def hashes(self) -> dict:
        return {m.content_hash: m for m in self.all(include_inactive=True)}

    def write(self, m: Memory) -> pathlib.Path:
        m.validate()
        self.mem_dir.mkdir(parents=True, exist_ok=True)
        p = self.mem_dir / f"{m.id}.md"
        p.write_text(m.to_text(), encoding="utf-8")
        return p

    def set_status(self, mid: str, status: str, note: str = "") -> Memory:
        m = self.by_id(mid)
        if m is None:
            raise MemoryError(f"no such memory {mid}")
        m.status = status
        if note:
            m.extra["note"] = note
        self.write(m)
        return m

    # -------------------------------------------------------------- staging
    def stage(self, memories, message="jedimem: capture") -> str:
        files = {f"{STAGE_PREFIX}{m.id}.md": m.to_text() for m in memories}
        if not files:
            return ""
        return repo.stage_files(files, message, ref=self.ref, cwd=self.root)

    def pending(self) -> list:
        out = []
        for path in repo.staged_files(ref=self.ref, cwd=self.root, prefix=STAGE_PREFIX):
            try:
                out.append(Memory.from_text(
                    repo.staged_content(path, ref=self.ref, cwd=self.root) + "\n"))
            except MemoryError:
                continue
        return out

    def clear_pending(self, ids, message="jedimem: resolve pending"):
        paths = [f"{STAGE_PREFIX}{i}.md" for i in ids]
        if paths:
            repo.drop_staged(paths, ref=self.ref, cwd=self.root, message=message)

    def promote(self, ids, status="active") -> list:
        """Move staged candidates into committed files (working tree write)."""
        pend = {m.id: m for m in self.pending()}
        written = []
        for i in ids:
            m = pend.get(i)
            if m is None:
                continue
            m.status = status
            self.write(m)
            written.append(m)
        self.clear_pending(ids, message=f"jedimem: promote {len(written)} memory(ies)")
        return written

File: src/store.rs
//! Where memories live: the staging ref, and the committed files.

use crate::memory::{Memory, MemoryError};
use crate::repo;
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

pub const MEM_DIR: &str = ".jedimem/memories";
pub const STAGE_PREFIX: &str = "pending/";

pub struct Store {
    pub root: PathBuf,
    pub git_ref: String,
}

impl Store {
    pub fn new(root: &Path, git_ref: &str) -> Store {
        Store {
            root: root.to_path_buf(),
            git_ref: git_ref.to_string(),
        }
    }

    pub fn mem_dir(&self) -> PathBuf {
        self.root.join(MEM_DIR)
    }

    // -------------------------------------------------------- committed
    pub fn all(&self, include_inactive: bool) -> Result<Vec<Memory>, MemoryError> {
        let dir = self.mem_dir();
        if !dir.is_dir() {
            return Ok(vec![]);
        }
        let mut paths: Vec<PathBuf> = std::fs::read_dir(&dir)
            .map_err(|e| MemoryError(e.to_string()))?
            .filter_map(|e| e.ok().map(|e| e.path()))
            .filter(|p| p.extension().map(|x| x == "md").unwrap_or(false))
            .collect();
        paths.sort();
        let mut out = Vec::new();
        for p in paths {
            let m = Memory::load(&p)?;
            if include_inactive || m.status == "active" {
                out.push(m);
            }
        }
        Ok(out)
    }

    pub fn by_id(&self, id: &str) -> Option<Memory> {
        let p = self.mem_dir().join(format!("{}.md", id));
        if p.exists() {
            Memory::load(&p).ok()
        } else {
            None
        }
    }

    pub fn hashes(&self) -> BTreeMap<String, Memory> {
        self.all(true)
            .unwrap_or_default()
            .into_iter()
            .map(|m| (m.content_hash(), m))
            .collect()
    }

    pub fn write(&self, m: &Memory) -> Result<PathBuf, MemoryError> {
        m.validate(None)?;
        let dir = self.mem_dir();
        std::fs::create_dir_all(&dir).map_err(|e| MemoryError(e.to_string()))?;
        let p = dir.join(format!("{}.md", m.id));
        std::fs::write(&p, m.to_text()).map_err(|e| MemoryError(e.to_string()))?;
        Ok(p)
    }

    pub fn set_status(&self, id: &str, status: &str, note: &str) -> Result<Memory, MemoryError> {
        let mut m = self
            .by_id(id)
            .ok_or_else(|| MemoryError(format!("no such memory {}", id)))?;
        m.status = status.to_string();
        if !note.is_empty() {
            m.extra.insert("note".into(), note.to_string());
        }
        self.write(&m)?;
        Ok(m)
    }

    // ---------------------------------------------------------- staging
    pub fn stage(&self, memories: &[Memory], message: &str) -> Result<String, repo::GitError> {
        if memories.is_empty() {
            return Ok(String::new());
        }
        let mut files = BTreeMap::new();
        for m in memories {
            files.insert(format!("{}{}.md", STAGE_PREFIX, m.id), m.to_text());
        }
        repo::stage_files(&files, message, &self.git_ref, Some(&self.root))
    }

    pub fn pending(&self) -> Vec<Memory> {
        let mut out = Vec::new();
        for path in repo::staged_files(&self.git_ref, Some(&self.root), STAGE_PREFIX) {
            if let Ok(text) = repo::staged_content(&path, &self.git_ref, Some(&self.root)) {
                if let Ok(m) = Memory::from_text(&format!("{}\n", text)) {
                    out.push(m);
                }
            }
        }
        out
    }

    pub fn clear_pending(&self, ids: &[String], message: &str) -> Result<(), repo::GitError> {
        if ids.is_empty() {
            return Ok(());
        }
        let paths: Vec<String> = ids
            .iter()
            .map(|i| format!("{}{}.md", STAGE_PREFIX, i))
            .collect();
        repo::drop_staged(&paths, &self.git_ref, Some(&self.root), message)?;
        Ok(())
    }

    /// Move staged candidates into committed files (a working-tree write, and
    /// therefore only ever on an explicit human action).
    pub fn promote(&self, ids: &[String], status: &str) -> Result<Vec<Memory>, MemoryError> {
        let pending: BTreeMap<String, Memory> = self
            .pending()
            .into_iter()
            .map(|m| (m.id.clone(), m))
            .collect();
        let mut written = Vec::new();
        for id in ids {
            if let Some(m) = pending.get(id) {
                let mut m = m.clone();
                m.status = status.to_string();
                self.write(&m)?;
                written.push(m);
            }
        }
        let msg = format!("jedimem: promote {} memory(ies)", written.len());
        self.clear_pending(ids, &msg)
            .map_err(|e| MemoryError(e.to_string()))?;
        Ok(written)
    }
}
